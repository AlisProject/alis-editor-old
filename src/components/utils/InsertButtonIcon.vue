<template>
  <div class="editor-insert-button__icon" :class="{ 'is-active': active }">
    <div class="editor-insert-button__iconContent" v-html="src" />
  </div>
</template>

<script lang="ts">
export default {
  props: {
    active: Boolean,
    src: String
  }
}
</script>

<style>
.editor-insert-button__icon {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: all 0.15s ease-out;
}

.editor-insert-button__iconContent {
  transform: scale(0.6);
}

.editor-insert-button__icon .fill-target {
  fill: #5e5e5e;
}

.editor-insert-button__icon.is-active {
  background: #0086cc;
}

.editor-insert-button__icon.is-active .fill-target {
  fill: #fff;
}
</style>
