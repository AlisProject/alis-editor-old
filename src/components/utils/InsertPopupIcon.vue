<template>
  <div class="editor-insert-popup__icon" :class="{ 'is-active': active }">
    <div class="editor-insert-popup__iconContent" v-html="src" />
  </div>
</template>

<script lang="ts">
export default {
  props: {
    active: Boolean,
    src: String
  }
}
</script>

<style scoped>
.editor-insert-popup__icon {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: all 0.15s ease-out;
  cursor: pointer;
}

.editor-insert-popup__iconContent {
  transform: scale(1);
}

.editor-insert-popup__icon .fill-target {
  fill: #fff;
}

.editor-insert-popup__icon.is-active .fill-target,
.editor-insert-popup__icon:hover .fill-target {
  fill: #848cd9;
}
</style>
