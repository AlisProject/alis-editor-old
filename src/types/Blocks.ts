export enum BlockType {
  Rule = 'Rule',
  Text = 'Text',
  Paragraph = 'Paragraph',
  Image = 'Image',
  Quote = 'Quote',
  Heading = 'Heading',
  Embed = 'Embed'
}

export interface Block {
  id: string
  type: BlockType
  payload?: any
  children?: Block[]
}

export interface RuleBlock extends Block {
  type: BlockType.Rule
}

export interface EmbedBlock extends Block {
  type: BlockType.Embed,
  payload: {
    src: string
  }
}

export interface ParagraphBlock extends Block {
  type: BlockType.Paragraph
  payload: {
    body: string
  }
}

export interface TextBlock extends Block {
  type: BlockType.Text
  payload: {
    body: string
  }
}

export interface ImageBlock extends Block {
  type: BlockType.Image
  payload: {
    src: string
    align?: string
  }
  children: Block[]
}

export interface HeadingBlock extends Block {
  type: BlockType.Heading
  payload: {
    size: 'h2' | 'h3'
    body: string
  }
  children: Block[]
}

export interface QuoteBlock extends Block {
  type: BlockType.Quote
  children: Block[]
}

/*
「詳細は [こちら](https://example.com) から確認できます」を再現する場合
[
  {
    type: 'Paragraph',
    children: [
      {
        type: 'Text',
        payload: {
          body: '詳細は'
        }
      },
      {
        type: 'Link',
        payload: {
          href: 'https://example.com'
        },
        children: [
          {
            type: 'Text',
            payload: {
              body: 'こちら'
            }
          }
        ]
      },
      {
        type: 'Text',
        payload: {
          body: 'から確認できます'
        }
      }
    ]
  }
]

*/
