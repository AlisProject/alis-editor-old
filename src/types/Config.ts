export interface AlisEditorConfig {
  articleId: string
  preview: boolean
  iframelyApikey: string
  uploadEndpoint: string
}
