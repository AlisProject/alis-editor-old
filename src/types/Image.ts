export interface ImageUploadRequest {
  article_image: string
}

export interface ImageUploadResponse {
  image_url: string
}
