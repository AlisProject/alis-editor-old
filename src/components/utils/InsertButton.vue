<template>
  <div
    class="insert-button"
  >
    <div
      class="insert-button__toggle"
      @click="toggleIsOpen"
      :class="{
        'is-open': isOpen
      }"
    >
      <span>+</span>
    </div>
    <ul class="insert-button__list" v-if="isOpen">
      <li class="insert-button__listItem" @click="dispatchUpload">
        画像
      </li>
      <li class="insert-button__listItem" @click="appendRule">
        罫線
      </li>
      <li class="insert-button__listItem" @click="appendQuote">
        引用
      </li>
      <li class="insert-button__listItem" @click="appendHeading">
        大見出し
      </li>
    </ul>
    <input type="file" @change="handleUpload" />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { BlockType } from '../../types/Blocks'

export default Vue.extend({
  data() {
    return {
      isOpen: false
    }
  },
  methods: {
    toggleIsOpen() {
      this.isOpen = !this.isOpen
    },
    dispatchUpload() {
      ;(this.$el.querySelector('[type="file"]') as any).click()
    },
    append(type: BlockType) {
      setTimeout(() => {
        this.isOpen = false
        this.$emit('append', type)
        this.$emit('disable')
      }, 100)
    },
    appendParagraph() {
      this.isOpen = false
      this.$emit('append', BlockType.Paragraph)
      this.$emit('disable')
    },
    appendRule() {
      this.append(BlockType.Rule)
    },
    appendQuote() {
      this.isOpen = false
      this.$emit('append', BlockType.Quote)
      this.$emit('disable')
    },
    appendHeading() {
      this.isOpen = false
      this.$emit('append', BlockType.Heading)
      this.$emit('disable')
    },
    handleUpload(event: Event) {
      this.isOpen = false
      this.$emit('disable')
      this.$emit('upload', event)
    },
    handleClickPublish() {
      this.$emit('publish')
    }
  }
})
</script>

<style scoped>
.insert-button {
  position: absolute;
  left: -40px;
  top: -1px;
  margin: 8px;
  cursor: pointer;
  z-index: 1000000000000000000;
}

.insert-button .insert-button__toggle {
  width: 30px;
  height: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
  border: solid 1px #ddd;
  color: #ddd;
  border-radius: 50%;
  font-weight: bold;
  padding-left: 1px;
  padding-bottom: 1px;
  transition: all 0.05s linear;
  transform: rotate(0deg);
  font-size: 25px;
  font-weight: normal;
  font-family: 'Yu Gothic', YuGothic;
  box-shadow: 0 0 10px 0 hsla(0, 0%, 57%, 0.5);
}

.insert-button__toggle span {
  pointer-events: event;
  user-select: none;
}

.insert-button .insert-button__toggle.is-open {
  transform: rotate(45deg);
}

.insert-button .insert-button__list {
  margin: 4px 0 0px 2px;
  padding: 0;
  color: #fff;
  border-radius: 4px;
  overflow: hidden;
  background: rgba(0, 0, 0, 0.9);
  display: flex;
  list-style-type: none;
  font-size: 1.4rem;
}

.insert-button__list .insert-button__listItem {
  padding: 4px 16px;
}

.insert-button__list .insert-button__listItem:hover {
  background: rgba(255, 255, 255, 0.2);
}

.insert-button__list .insert-button__listItem + .insert-button__listItem {
  border-left: solid 1px #fff;
}

input {
  display: none;
}
</style>
