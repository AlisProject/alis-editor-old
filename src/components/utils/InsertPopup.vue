<template>
  <div>
    <div
      class="insert-popup"
      v-if="activeRoot"
      :class="{ 'is-active': isActive }"
      :style="{
        left: `${this.position.left}px`,
        top: `${this.position.top}px`
      }"
    >
      <button class="insert-popup__item" @click="execBold">
        <InsertPopupIcon :src="require('../../assets/icon-bold.svg.js')" />
      </button>
      <button class="insert-popup__item" @click="execItalic">
        <InsertPopupIcon :src="require('../../assets/icon-italic.svg.js')" />
      </button>
      <button class="insert-popup__item" @click="execQuote">
        <InsertPopupIcon :src="require('../../assets/icon-quote.svg.js')" />
      </button>
      <button class="insert-popup__item" @click="execHeading">
        <InsertPopupIcon :src="require('../../assets/icon-h2.svg.js')" />
      </button>
      <!--
        <button class="insert-popup__item" @click="execSubHeading">
          <InsertPopupIcon :src="require('../../assets/icon-h3.svg.js')" />
        </button>
      -->
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import InsertPopupIcon from './InsertPopupIcon.vue'
import { BlockType } from '../../types/Blocks'

export default Vue.extend({
  props: {
    activeRoot: Object
  },
  data(): any {
    return {
      isOpen: false,
      isActive: false,
      el: null,
      position: {
        left: 0.0,
        top: 0.0
      }
    }
  },
  components: {
    InsertPopupIcon
  },
  mounted() {
    document.addEventListener('selectionchange', this.onSelectionChange)
  },
  beforeDestroy() {
    document.removeEventListener('selectionchange', this.onSelectionChange)
  },
  watch: {
    activeRoot(aR) {
      this.el = document.querySelector(`[data-block-id="${aR.id}"]`)
      this.position.left = (document.querySelector('#ALISEditor') as any).offsetWidth * 0.5 - 110
      this.position.top = this.el.getBoundingClientRect().top - 50 + window.pageYOffset
      if (aR.type === 'Paragraph') {
        this.isActive = false
      } else {
        this.isActive = true
      }
    }
  },
  methods: {
    onSelectionChange(event: Event) {
      if (this.activeRoot.type !== 'Paragraph') {
        return
      }
      const selection = window.getSelection()
      if (selection.isCollapsed) {
        this.isActive = false
        return
      }
      this.isActive = true
    },
    execBold() {
      document.execCommand('bold')
    },
    execItalic() {
      document.execCommand('italic')
    },
    execQuote() {
      this.$emit('replace', BlockType.Quote)
    },
    execHeading() {
      this.$emit('replace', BlockType.Heading)
    },
    execSubHeading() {
      this.$emit('replace', BlockType.Heading)
    }
  }
})
</script>

<style scoped>
.insert-popup {
  margin: 0;
  padding: 0;
  color: #fff;
  display: flex;
  list-style-type: none;
  font-size: 1.4rem;
  background: #41446a;
  width: 220px;
  position: absolute;
  z-index: 200;
  opacity: 0;
}

.insert-popup.is-active {
  animation: anim1 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275) forwards;
}

@keyframes anim1 {
  0% {
    transform: translateY(10px);
    opacity: 0;
  }
  100% {
    transform: translateY(0px);
    opacity: 1;
  }
}

.insert-popup::after {
  position: absolute;
  left: calc(50% - 10px);
  bottom: -10px;
  width: 20px;
  height: 20px;
  background: #41446a;
  transform: rotate(45deg);
  content: '';
  display: block;
}

.insert-popup .insert-popup__item {
  margin-left: 8px;
  width: 50px;
  height: 50px;
  display: flex;
  align-items: center;
  justify-content: center;
  color: #fff;
  font-weight: bold;
  padding-left: 1px;
  padding-bottom: 1px;
  transition: all 0.05s linear;
  transform: rotate(0deg);
  font-size: 25px;
  font-weight: normal;
  font-family: 'Yu Gothic', YuGothic;
  appearance: none;
  -webkit-appearance: none;
  background: transparent;
  border: none;
  outline: none;
}

input {
  display: none;
}
</style>
