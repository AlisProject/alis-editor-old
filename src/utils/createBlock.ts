import uuid from 'uuid/v4'
import { Block, BlockType } from '../types/Blocks'

export function isContentEditableBlock(block: Block) {
  return isContentEditableBlockType(block.type)
}

export function isContentEditableBlockType(type: BlockType) {
  return [BlockType.Paragraph].includes(type)
}

export function createBlock(type: BlockType, extend?: any): Block {
  extend = extend || {}
  return Object.assign({ id: uuid() }, createBlockTemplate(type), extend)
}

function createBlockTemplate(type: BlockType) {
  switch (type) {
    case BlockType.Embed: {
      return {
        type: BlockType.Embed,
        payload: {
          src: ''
        }
      }
    }
    case BlockType.Paragraph: {
      return {
        type: BlockType.Paragraph,
        payload: {
          body: '<p><br></p>'
        }
      }
    }
    case BlockType.Image: {
      return {
        type: BlockType.Image,
        payload: {
          src: '',
          align: 'center',
          caption: ''
        }
      }
    }
    case BlockType.Rule: {
      return {
        type: BlockType.Rule
      }
    }
  }
}
