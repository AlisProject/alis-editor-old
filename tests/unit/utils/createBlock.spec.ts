import { BlockType } from '../../../src/types/Blocks'
import { createBlock, isContentEditableBlock } from '../../../src/utils/createBlock'

describe('utils/createBlock.ts', () => {
  describe('isContentEditableBlock', () => {
    test('valid', () => {
      expect(isContentEditableBlock(createBlock(BlockType.Paragraph))).toBe(true)
    })
    test('invalid', () => {
      expect(isContentEditableBlock(createBlock(BlockType.Rule))).toBe(false)
    })
  })
  Object.entries(BlockType).forEach(([K, V]) => {
    test(K, () => {
      expect(
        Object.assign({}, createBlock(V, {}), {
          id: 'VVVV-XXXX-YYYY-ZZZZ'
        })
      ).toMatchSnapshot()
    })
  })
})
