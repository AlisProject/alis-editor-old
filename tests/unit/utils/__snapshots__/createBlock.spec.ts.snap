// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/createBlock.ts Blank-default 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
}
`;

exports[`utils/createBlock.ts Blank-extend 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
}
`;

exports[`utils/createBlock.ts Embed-default 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "payload": Object {
    "src": "",
  },
  "type": "Embed",
}
`;

exports[`utils/createBlock.ts Embed-extend 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "payload": Object {
    "src": "",
  },
  "type": "Embed",
}
`;

exports[`utils/createBlock.ts Image-default 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "payload": Object {
    "align": "center",
    "caption": "",
    "src": "",
  },
  "type": "Image",
}
`;

exports[`utils/createBlock.ts Image-extend 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "payload": Object {
    "align": "center",
    "caption": "",
    "src": "",
  },
  "type": "Image",
}
`;

exports[`utils/createBlock.ts Paragraph-default 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "payload": Object {
    "body": <p></p>,
  },
  "type": "Paragraph",
}
`;

exports[`utils/createBlock.ts Paragraph-extend 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "payload": Object {
    "body": <p></p>,
  },
  "type": "Paragraph",
}
`;

exports[`utils/createBlock.ts Rule-default 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "type": "Rule",
}
`;

exports[`utils/createBlock.ts Rule-extend 1`] = `
Object {
  "id": "VVVV-XXXX-YYYY-ZZZZ",
  "type": "Rule",
}
`;
